### PR TITLE
change gpapply to run with new database & dataframe

### PR DESCRIPTION
**Pull request by violet2016**

### CODE REVIEW

**Comment by HaozhouWang** [comment]
If output is `None`, we do not need to `create table`
`input_py_func_extra_args` is appended as py_func arguments

File: greenplumpython/gpapply.py [hunk 1]
```diff
@@ -13,94 +15,59 @@ def randomStringType(stringLength=10):
...
-output list("a", "int8", "b", "float")
-index "col1"
-"""
+def pythonExec(df, funcName, typeName, output_tbl_name, extra_args):
+    func_type_name = []
+
+    internal_select = []
+    for i, col in enumerate(df.table_metadata.signature):
+        for j, column in enumerate(col):
+            internal_select.append("array_agg(" + column + ") AS " + column)
```
Comment (HaozhouWang) [inline greenplumpython/gpapply.py:24]: No array_agg needed.

File: greenplumpython/gpapply.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import inspect
-import psycopg2
 import random
 import string
+import numpy as np
+
+numpyDTypeToGPDBType = {np.dtype('int64'): "int", np.dtype('object'): "text", np.dtype("datetime64[ns]"):"datetime"}
```
Comment (wengyanqing) [inline greenplumpython/gpapply.py:6]: Any reference for this global variable ?

File: greenplumpython/gpapply.py [hunk 3]
```diff
@@ -13,94 +15,59 @@ def randomStringType(stringLength=10):
     letters = string.ascii_lowercase
     return "tmp_type"+''.join(random.choice(letters) for i in range(stringLength))
 
-"""
-list("a", "int4", "b", "int4")
-output list("a", "int8", "b", "float")
-index "col1"
-"""
+def pythonExec(df, funcName, typeName, output_tbl_name, extra_args):
```
Comment (wengyanqing) [inline greenplumpython/gpapply.py:18]: How to handle extra_args?

File: greenplumpython/gpapply.py [hunk 4]
```diff
@@ -13,94 +15,59 @@ def randomStringType(stringLength=10):
...
-                tbl_internal_select = tbl_internal_select + ", array_agg(" + input[i] + ") AS " + input[i]    
-    
-    
     select_func = "CREATE TABLE " + output_tbl_name + " AS \n" \
         + "WITH gpdbtmpa AS ( \n" \
-        + "SELECT (" + funcName + "(" + func_type_name + ")) AS gpdbtmpb FROM (SELECT " \
-        + tbl_internal_select + " FROM " + tbl_name + " GROUP BY " + index + ") tmptbl \n ) \n" \
+        + "SELECT (" + funcName + "(" + ",".join(func_type_name) + ")) AS gpdbtmpb FROM (SELECT " \
+        + ",".join(internal_select) + " FROM " + df.table_metadata.name + ") tmptbl \n ) \n" \
         + "SELECT (gpdbtmpb::" + typeName + ").* FROM gpdbtmpa DISTRIBUTED RANDOMLY;"
```
Comment (wengyanqing) [inline greenplumpython/gpapply.py:32]: It's better to make result table distribution policy configurable.

File: greenplumpython/gpapply.py [hunk 5]
```diff
@@ -13,94 +15,59 @@ def randomStringType(stringLength=10):
...
-    cur.execute(drop_sql)
-    output_select_sql = "select * from " + output_tbl_name + ";"
-    cur.execute(output_select_sql)
-    res = cur.fetchall()
-    conn.commit()
-    return res
+    if clear_existing:
+        drop_table_sql = "drop table if exists %s;" % output.name
+        db.execute(drop_table_sql)
+    db.execute(create_type_sql)
```
Comment (wengyanqing) [inline greenplumpython/gpapply.py:69]: It's better to make these sql execution in transaction block.

File: greenplumpython/tests/test_gpapply.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    data = GPDatabase(db_conn)
+    table = data.get_table("basic", "public")
+    output_col = [{"a":"int4"}]
+    output = GPTableMetadata("basic_output", output_col, 'randomly')
+    index = "a"
+    gpApply(table, recsum, data, output)
+    res = data.execute_query("select * from basic_output")
+    assert res.iat[0,0] == 7
+
+def test_gpapply_case2(db_conn):
```
Comment (wengyanqing) [inline greenplumpython/tests/test_gpapply.py:43]: It's better to add some test cases which will throw exception.

File: greenplumpython/gpapply.py (hunk 3 above)
Comment (violet2016) [reply greenplumpython/gpapply.py:18]: It's not handled yet, we need another story for this arg

File: greenplumpython/gpapply.py (hunk 5 above)
Comment (violet2016) [reply greenplumpython/gpapply.py:69]: we already have a story for this #172082257

File: greenplumpython/gpapply.py (hunk 2 above)
Comment (violet2016) [reply greenplumpython/gpapply.py:6]: It's useless now, I'll delete it

**Pull request merged**